test/files/run/t1980.scala

```diff
@@ -0,0 +1,43 @@
+class LazyList[+A](expr: => LazyList.Evaluated[A]) {
+  def #:: [B >: A](elem: => B): LazyList[B] = new LazyList(Some((elem, this)))
+  def ##:: [B >: A](elem: B): LazyList[B] = new LazyList(Some((elem, this)))
+  def force: Unit = expr.foreach(_._2.force)
+}
+
+object LazyList {
+  type Evaluated[+A] = Option[(A, LazyList[A])]
+  object Empty extends LazyList[Nothing](None)
+  def empty[A]: LazyList[A] = Empty
+}
+
+object Test extends App {
+  def foo(i: Int) = { println("foo "+i); i }
+  def bar(i: Int) = { println("bar "+i); i }
+  println("defining")
+  val xs1 = foo(1) ##:: foo(2) ##:: foo(3) ##:: LazyList.empty
+  val xs2 = bar(1) #:: bar(2) #:: bar(3) #:: LazyList.empty
+  println("forcing")
+  xs1.force
+  xs2.force
+
+  {
+    class C { def f_:(x: => Int)(implicit y: Int) = 0 }
+    val c = new C
+    implicit val i = 1
+    def k = { println("hi"); 1 }
+    c.f_:(k)
+    k f_: c
```

Comment:
maybe also test lazy evaluation here (and in all tests below), just to make sure.